doc/specs/#140 - ZIP Support.md

```diff
@@ -0,0 +1,90 @@
+
+> Since we are utilizing Windows Shell APIs, we need to ensure that we are not invoking a UI during the extraction. This can be done by [setting the operation flag to not display any UI](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileoperation-setoperationflags). 
+
+> During implementation, we will also need to ensure that this process can work under SYSTEM context.
+
+## ZIP Threat Detection
+ZIP and other archive file types are known threat vectors for malware in the form of ZIP compression bombs. Two of the most common types of compression bombs are multi-layered (recursive) and single-layered (non-recursive). ZIP bombs rely on a repetition of identical files that have extremely large compression ratios. 
+
+1. **Multi-layered (recursive)**:
+A ZIP file containing multiple layers of nested ZIP files that achieve high compression ratios when decompressed recursively. This technique is often used to bypass compression ratio checks performed by ZIP parsers for a single layer.
+
+2. **Single-Layered (non-recursive)**
+A ZIP bomb that expands fully after a single round of decompression. The extremely high compression ratio of the ZIP bomb is achieved by overlapping files within the ZIP container. 
+
+In order to protect our users from these possible threats, we will need to implement several basic checks to verify whether it is safe to proceed with the contents of the ZIP archive file.
+
+>Compression Ratio is defined as the sum of the sizes of all the items contained within a ZIP file, divided by the size of the ZIP file.
+
+The check will be as follows:
+
+1. For any archive file, we will start by using Windows Shell APIs which treat ZIP files as a Folder object that can be utilized for viewing the contents (we are not extracting to a specified location, yet). We will iterate through each item and determine the cumulative sum of the sizes of all items contained in the zip and compare that value with the size of the ZIP file. **The check will verify that the compression ratio is no greater than 5:1 (20%)**. Any file that has a larger compression ratio (<20%) will be deemed as "suspicious" and a warning will be displayed to the user. Lossless compression of common data files tend to achieve compression ratios of ~2:1 (50%). Any extraction or copying of files will not take place until this check is passed.
+
+2. If a ZIP archive file contains nested archive files, we will support decompressing up to a maximum of 3 layers (2 additional layers including the initial ZIP file). 

```

Comment:
Is this recursive decompressing talking only about the threat detection, or is it part of the installation?

If part of the installation, how do you intend to use decompressing multiple nested archives?
For example:
Would I be able to specify in the manifest that I want to run the installer `/archive.zip/setup.exe`? Where would you decompress the nested archives to? What if I gave you a zip containing `setup.exe` and `data.zip`, and `setup.exe` used the compressed `data.zip`; would you still decompress it?

Reply:
Okay. Just to be annoying bringing up edge cases...
Imagine you have `foo.zip` contains `bar.zip` which contains `setup.exe`. But `foo.zip` also contains `bar/setup.exe` (`setup.exe` inside a directory called `bar`). Both would be decompressed to `bar/setup.exe`. What do we do then?

Reply:
I think setup.exe with data.zip is a valid scenario. Maybe we should only unzip the top layer as default and add some manifest entries to let us know we should unzip additional layers?

Reply:
> Okay. Just to be annoying bringing up edge cases... Imagine you have `foo.zip` contains `bar.zip` which contains `setup.exe`. But `foo.zip` also contains `bar/setup.exe` (`setup.exe` inside a directory called `bar`). Both would be decompressed to `bar/setup.exe`. What do we do then?

Not annoying :) I like Yao's suggestion to include a manifest entry like `fullDecompression = true` to indicate whether we decompress additional layers. If there is already an existing folder with the same name, maybe the decompressed folder would append a string to the end of the name like `"bar (1)"` to signify that there is a duplicate folder. This might be slighly confusing but my hope is that there wouldn't be any nested zips that conflict with nested folders that contain the same files.

Reply:
Do we have a use case for extracting nested archives?  It seems more likely to break things than help, in the event that a data.zip file exists.  I would say that until a concrete example exists, we should start with only extracting the top level archive.

Reply:
Added explanation to specify that initial implementation will only extract the top-level archive to support the scenarios stated above